### PR TITLE
Introduce an UpdatePolicyManager (QC-342)

### PR DESCRIPTION
**Pull request by Barthelemy**
Policies and revisions were spread over `Check` and `CheckRunner` making it difficult to comprehend and to reuse (e.g. in the future aggregators). 

This PR proposes an extraction of these concepts in a separate class called `PolicyManager`. The `Check` class is unaware of the notion of revisions. It still holds its own configuration and thus the policy definition. As a consequence the PolicyManager is indirectly influenced by the way the `Check` class overwrites the policy name in certain cases. This is something we should reconsider in the future. 

I have kept the use of lambdas to implement the actual logic of the different types of policies. Again, this is something we might want to reconsider in the future. The fact that there is a `policyHelper` parameter necessary to hold a state of one of the type of policies is a sign that we might want to move away from simple lambdas. A more traditional OO approach with a common interface and a set of classes implementing the actual policy function could be a solution. 

All in all it is not my dream refactoring but it achieves the extraction of the code related to policies and revisions from `Check` and `CheckRunner` and improves the separation of concerns.

### CODE REVIEW

**Comment by Barthelemy** [comment]
No idea what is going on with the CI.

**Comment by Barthelemy** [comment]
@ktf o2-dataflow  is still failing

**Comment by ktf** [comment]
restarted

File: Framework/test/testPolicyManager.cxx [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+// or submit itself to any jurisdiction.
+
+///
+/// \file   testPolicyManager.cxx
+/// \author Barthelemy von Haller
+///
+
+#include "QualityControl/PolicyManager.h"
+#include <DataSampling/DataSampling.h>
+using namespace o2::utilities;
```
Comment (knopers8) [inline Framework/test/testPolicyManager.cxx:18]: Not needed

**Review by knopers8 — CHANGES_REQUESTED** [review]
Thanks, this looks much better already!

I would propose to change `Policy` to `UpdatePolicy`, since the former is too general IMO. Likewise, `PolicyManager` to `UpdatePolicyManager`

Could you also change the PR name to something more accurate? After all, Policies didn't become classes yet.

In later refactoring, I also thought about making UpdatePolicies return object names which should be checked. That should clean up the `Check::check()` method, which is a monster now. This probably is for later though.

**Comment by Barthelemy** [comment]
Thanks for the great review. I'll update the PR accordingly. 

@MichaelLettrich It seems that the mac builder is broken again.

**Comment by knopers8** [comment]
I have tried manually if the OneEachSeparately policy still works - it seems it does.

File: Framework/src/PolicyManager.cxx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+     * Return true if any MOs were updated.
+     * Inner policy - used for `"MOs": "all"`
+     * Might return true even if MO is not used in Check
+     */
+
+    policy = []() {
+      // Expecting check of this policy only if any change
+      return true;
+    };
+
```
[the remark this answers is not in the document]
Comment (Barthelemy) [reply Framework/src/PolicyManager.cxx:135]: I left it in order not to touch the logic that pre-existed between the policies and the Check class. The Check updates the policy depending for example on the "all" parameter and sets it to this _OnGlobalAny. I would leave it for now and wait for the next round of refactoring.

[the remark this answers is not in the document]
Comment (knopers8) [reply Framework/src/PolicyManager.cxx:135]: ok

File: Framework/CMakeLists.txt [hunk 3]
```diff
@@ -65,7 +65,7 @@ add_library(QualityControl
             src/HistoProducer.cxx
             src/DataProducerExample.cxx
             src/MonitorObjectCollection.cxx
-            src/PolicyManager.cxx)
+  src/UpdatePolicyManager.cxx)
```
Comment (knopers8) [inline Framework/CMakeLists.txt:68]: Please fix the indentation

File: Framework/CMakeLists.txt [hunk 4]
```diff
@@ -121,7 +121,7 @@ add_root_dictionary(QualityControl
                             include/QualityControl/TrendingTask.h
                             include/QualityControl/Reductor.h
                             include/QualityControl/MonitorObjectCollection.h
-                            include/QualityControl/PolicyManager.h
+  include/QualityControl/UpdatePolicyManager.h
```
Comment (knopers8) [inline Framework/CMakeLists.txt:124]: Please fix the indentation

File: Framework/src/TaskRunner.cxx [hunk 5]
```diff
@@ -280,8 +280,7 @@ void TaskRunner::loadTopologyConfig()
 {
   auto taskConfigTree = getTaskConfigTree();
   auto policiesFilePath = mConfigFile->get<std::string>("dataSamplingPolicyFile", "");
-  ConfigurationInterface* config = policiesFilePath.empty() ? mConfigFile.get() : ConfigurationFactory::getConfiguration(policiesFilePath).get();
-  auto policiesTree = config->getRecursive("dataSamplingPolicies");
+  ConfigurationInterface* config = policiesFilePath.empty() ? mConfigFile.get() : ConfigurationFactory::getConfiguration(policiesFilePath).get(); // todo make sure that it is ok to use the internal pointer of the unique_ptr
```
Comment (knopers8) [inline Framework/src/TaskRunner.cxx:283]: nice catch, it looks dangerous indeed

**Review by knopers8 — APPROVED** [review]
Thanks for the fixes, please just fix the formatting in CMakeLists after CLion (i guess) destroyed it.

**Pull request merged**